### ISSUE TITLE
Before each?

### ISSUE THREAD

**Issue by SimonThormeyer** [evanwashere/mitata#37]
Hi, 
first of all, thanks a lot for this great tool.

I was wondering how I get `beforeEach` semantics for benchmark execution? Or is this feature currently missing in mitata?

My use case is something like this:

```typescript
import assert from "node:assert";
import { bench, run } from "mitata";

bench("list pop()", async function* () {
    // this initilization needs to be run for each bench execution:
    const list = [1, 2];  
    yield async () => {
        // object gets consumed inside bench
        const result = list.pop();
         // error :(
        assert(result === 2); 
    };
});

await run();
```

It would be great if there was a way to not only yield the actual function to be measured, but a `beforeEach` function along with it.

**Comment by evanwashere** [comment]
There are no loop/iteration hooks in mitata because internally it's not always 1:1 match and it directly affects the numbers in micro-benchmarks (sub 100-1000ns)

Instead you should do expensive part of preparation outside of loop and make a cheap copy or just create the array in loop (mimicking real application)

```
import { bench, run } from "mitata";

bench("list pop()", async function* () {
    const _list = [1, 2];  
    yield async () => {
        const list = _list.slice();
    };
});

await run();
```

**Comment by SimonThormeyer** [comment]
Thanks @evanwashere for your speedy response!

Unfortunately, **I cannot make cheap copies** in my use case. Sorry that I over-simplified. 

The only option I would have is to create upfront an array that contains all arrays that I'm going to need.
The downside here is that I have to know upfront how many lists I have to provide so that the bench always has one ready to consume. Having to guess that would be a quite big limitation (or does mitata provide an API to set/get the maximum number of executions?).

AFAICT the only thing I can do is
```typescript
import assert from "node:assert";
import { bench, run } from "mitata";

async function* generateLists(
    neededListsCount: number
): AsyncGenerator<number[]> {
    const lists = [];
    for (let i = 0; i < neededListsCount; i++) {
        lists.push([1, 2]);
    }

    for (const list of lists) {
        yield list;
    }
}

bench("list pop()", async function* () {
    // I have to guess this, right?
    const EXPECTED_BENCH_EXECUTION_COUNT = 9999999;
    const lists = generateLists(EXPECTED_BENCH_EXECUTION_COUNT);

    yield async () => {
        const { value: list } = await lists.next();
        const result = list.pop();
        assert(result === 2);
    };
});


await run();
```

Where having to guess the expected execution count is far from ideal.

Is not being able to make cheap copies of a state that is altered in the bench really such a special case that you wouldn't consider supporting an API like this? (If this is possible/compatible with your internals :))

```typescript
bench("list pop", async function* () {
    bench("list pop()", async function* () {
        yield {
            // Leave undefined if not needed.
            async before() {
                return [1, 2];
            },
            async bench(list: number[]) {
                // object gets consumed inside bench
                const result = list.pop();
                assert(result === 2);
            },
        };
    });
});
```

If you're not going to consider something like this, I will probably run 2 benches, A and B, where A contains both setup and the operation I'm actually interested in, and B contains just the setup, then I'll subtract the runtumes of B from A to get the result I need.
Then I will, however, not get all the nice visualizations of the result I need, which would be a pity.
Thanks again! 😄

**Comment by evanwashere** [comment]
hooks that allow observing loop behavior reduce timing precision so they will never be added but I came up with [computed parameters](https://github.com/evanwashere/mitata?tab=readme-ov-file#computed-parameters) to solve your problem (available from ^1.0.17)
